org.jboss.tools.vscode.java/src/org/jboss/tools/vscode/java/internal/handlers/CodeActionHandler.java

```diff
@@ -0,0 +1,148 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Red Hat Inc. - initial API and implementation
+ *******************************************************************************/
+package org.jboss.tools.vscode.java.internal.handlers;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.jdt.core.IBuffer;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.compiler.IProblem;
+import org.eclipse.jdt.core.dom.CompilationUnit;
+import org.eclipse.lsp4j.CodeActionParams;
+import org.eclipse.lsp4j.Command;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.TextEdit;
+import org.eclipse.lsp4j.WorkspaceEdit;
+import org.jboss.tools.vscode.java.internal.JDTUtils;
+import org.jboss.tools.vscode.java.internal.SharedASTProvider;
+import org.jboss.tools.vscode.java.internal.TextEditConverter;
+import org.jboss.tools.vscode.java.internal.corrections.DiagnosticsHelper;
+import org.jboss.tools.vscode.java.internal.corrections.UnusedCodeCorrections;
+
+/**
+ * @author Gorkem Ercan
+ *
+ */
+public class CodeActionHandler {
+
+	/**
+	 *
+	 */
+	private static final String COMMAND_ID_APPLY_EDIT = "java.apply.workspaceEdit";
+
+
+
+	/**
+	 * @param params
+	 * @return
+	 */
+	public List<Command> getCodeActionCommands(CodeActionParams params) {
+		final ICompilationUnit unit = JDTUtils.resolveCompilationUnit(params.getTextDocument().getUri());
+		List<Command> $ = new ArrayList<>();
+		params.getContext().getDiagnostics().stream()
+		.filter(this::hasCodeAssist)
+		.forEach(diagnostic-> {$.addAll(getCommandForDiagnostic(unit, params.getTextDocument().getUri(), diagnostic));});
+		return $;
+	}
+
+
+	private boolean hasCodeAssist(Diagnostic diagnostic){
+
+		final int problemId= getProblemId(diagnostic);
+		switch(problemId){
+		case IProblem.UnterminatedString:
+		case IProblem.UnusedImport:
+		case IProblem.DuplicateImport:
+		case IProblem.CannotImportPackage:
+		case IProblem.ConflictingImport:
+		case IProblem.ImportNotFound:
+			return true;
+		default:
+			return false;
+		}
+	}
+
+	private List<Command> getCommandForDiagnostic(ICompilationUnit unit, String uri,Diagnostic diagnostic){
+		final int problemId = getProblemId(diagnostic);
+		switch(problemId){
+		case IProblem.UnterminatedString:
+			String label = "Insert missing quote";
+			WorkspaceEdit edit = new WorkspaceEdit();
+			try {
+				int offset = DiagnosticsHelper.getEndOffset(unit, diagnostic);
+				int start = DiagnosticsHelper.getStartOffset(unit, diagnostic);
+				int pos = moveBack(offset, start, "\n\r", unit);
+				TextEdit te = new TextEdit(JDTUtils.toRange(unit, pos, 0), "\"");
```

Comment:
doesn't sound like a good idea to generate textedits for position 0, in case an error occurred during computation. Probably better to not return anything